### PR TITLE
Add maps quiz

### PR DESCRIPTION
**Pull request by mortum5**
#15 issue

### CODE REVIEW

File: challenge/files/11_maps.toml [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+[[learning_links]]
+title = "Go by examples: Maps"
+url = "https://gobyexample.com/maps"
+
+[[questions]]
+text = """Select the true statements and the line they refer to.
+"""
+type = "select_answers"
+answers = [
+    { text = "The map can be used without initialization" },
```
Comment (rusinikita) [inline challenge/files/11_maps.toml]: I think that first question overwhelms. We need to split it.

Screenshot 2024-08-11 at 15 47 57

### First question.

```go
func main() {
	var m map[int]bool
	
        val, ok := m[1]
	fmt.Println(val, " ", ok)
	
        for i := 0; i < 5; i++ {
		m[i] = i%2 == 0
	}

	fmt.Println(m)
```

What program prints - not compiles, panic, false false, false false and panic

### Fixed initialization questions

Replace code with new snippet with fixed map initialization. 

```go
func main() {
	var m map[int]bool
	fmt.Println(m)
	
        for i := 0; i < 5; i++ {
		m[i] = i%2 == 0
	}

	fmt.Println(m)
```

What program prints. Plus next question - why.

### random order question

```go
func main() {
	var m map[int]bool
	fmt.Println(m)

	val, ok := m[1]
	fmt.Println(val, ok)

	for i := 0; i < 5; i++ {
		m[i] = i%2 == 0
	}

	var toPrint []int
	for key := range m {
		toPrint = append(toPrint, key)
	}

	fmt.Println(toPrint)
}
```

What it prints. Next question - why

### clean question

I think we should skip it.

Comment (rusinikita) [reply challenge/files/11_maps.toml]: You can change co with question `new_code_snippet` field

Screenshot 2024-08-11 at 16 36 47

File: challenge/files/11_maps.toml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+[[questions]]
+text = """Which of the following statements about the hash function used in Go maps is true?
+
+"""
+type = "select_answers"
+answers = [
+    { text = "Go uses a simple modulo operation to compute hash values for keys" },
+    { text = "The hash function is deterministic and always produces the same hash for the same input across different runs" },
+    { text = "The hash function includes a per-map random seed to protect against certain types of attacks", code_line_ranges = [] },
```
Comment (rusinikita) [inline challenge/files/11_maps.toml]: Let's specify the attack name and add proof in the learning links. I don't know about it and want to read)

**Review by rusinikita — CHANGES_REQUESTED** [review]

File: challenge/files/11_maps.toml (hunk 2 above)
Comment (mortum5) [reply challenge/files/11_maps.toml]: it's not like there's some famous attack using this. I just took it from what I read on a blog. Here's a link, but it's not an official document. I can remove this question.

[Link to Hackernoon](https://hackernoon.com/some-insights-on-maps-in-golang-rm5v3ywh#:~:text=I%20think%2C%20the%20idea%20behind%20this%20randomization%20is%20to%20avoid%20potential%20vulnerabilities%20when%20malicious%20actor%20will%20flood%20hashmap%20with%20collisions%20and%20to%20prevent%20any%20attempts%20to%20rely%20on%20map%20iteration%20order.%20Map%20is%20iterable%2C%20but%20not%20ordered.%20Even%20every%20new%20iterator%20of%20the%20same%20map%20is%20randomized)

File: challenge/files/11_maps.toml [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+	for i := 0; i < 5; i++ {
+		m[i] = i%2 == 0
+	}
+
+	fmt.Println(m)
+}
+"""
+learning_advise = """
+“Learning is never done without error, and defeat ” – Vladimir Lenin.
```
Comment (rusinikita) [inline challenge/files/11_maps.toml]: Test data. Let's write something useful

File: challenge/files/11_maps.toml [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    { text = "[3 4 0 1 2]"},
+    { text = "[4 3 2 1 0]"},
+]
+
+[[questions]]
+text = """Explain the output of the program
+"""
+type = "select_answers"
+answers = [
+    { text = "Iteration over map keys is randomized between every for range call", code_line_ranges = [] },
```
Comment (rusinikita) [inline challenge/files/11_maps.toml]: Can't understand "between every for range call"

File: challenge/files/11_maps.toml [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+type = "select_answers"
+answers = [
+    { text = "Each bucket can store exactly one key-value pair" },
+    { text = "Buckets use a linked list to store multiple key-value pairs that hash to the same bucket" },
+    { text = "Buckets store key-value pairs in an array and resolve collisions using open addressing" },
+    { text = "Buckets store key-value pairs in an array and use a linked list to handle collisions", code_line_ranges = [] },
+]
+
+[[questions]]
+text = """What happens when the number of entries in a Go map exceeds a certain threshold?
```
Comment (rusinikita) [inline challenge/files/11_maps.toml]: Don't understand "certain threshold"

File: challenge/files/11_maps.toml [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+answers = [
+    { text = "The map allocates a new array with double the capacity and rehashes all keys" },
+    { text = "The map remains the same size but uses a more sophisticated hash function" },
+    { text = "The map splits existing buckets and rehashes only a portion of the keys", code_line_ranges = [] },
+    { text = "The map creates a second, independent map and balances the keys between the two maps" },
+]
+
+[[questions]]
+text = """The load factor of a map is a critical concept in understanding its efficiency. 
+Which of the following best describes the load factor in the context of Go maps?
```
Comment (rusinikita) [inline challenge/files/11_maps.toml:143]: Maybe this question needs to be before "What happens when"

File: challenge/files/11_maps.toml (hunk 2 above)
Comment (rusinikita) [reply challenge/files/11_maps.toml]: Let's specify "attack when malicious actor will flood hashmap with collisions"

And let's add link in useful links

Once, I was asked about bucket overflow (or something like that) when the attacker tries to find keys to overflow only a single bucket it makes service significantly slower.

File: challenge/files/11_maps.toml [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+"""
+type = "select_answers"
+answers = [
+    { text = "Go uses a simple modulo operation to compute hash values for keys" },
+    { text = "The hash function is deterministic and always produces the same hash for the same input across different runs" },
+    { text = "The hash function includes a per-map random seed to protect against certain types of attacks", code_line_ranges = [] },
+    { text = "Go maps do not use a hash function; they rely on direct indexing" },
+]
+
+
```
Comment (rusinikita) [inline challenge/files/11_maps.toml:204]: We need to add the question "What happens when we need to add new value, but the bucket is full"

**Review by rusinikita — CHANGES_REQUESTED** [review]

**Review by rusinikita — APPROVED** [review]

**Pull request merged**